### PR TITLE
Add `RotationGenerator` and its subtypes

### PR DESCRIPTION
**Pull request by hyrodium**
This PR solves #30, #31, ~~#126 and #128.~~ (2021/11/20 EDIT: quaternion conversions will be supported in the other PR.)

```julia
julia> using Rotations

julia> m = rand(3,3)
3×3 Matrix{Float64}:
 0.992789  0.0478161  0.639771
 0.826576  0.89928    0.368333
 0.100041  0.986343   0.174436

julia> s = InfinitesimalRotMatrix{3}(m - m')
3×3 InfinitesimalRotMatrix{3, Float64, 9} with indices SOneTo(3)×SOneTo(3):
  0.0      -0.77876   0.53973
  0.77876   0.0      -0.61801
 -0.53973   0.61801   0.0

julia> exp(s)
3×3 RotationVec{Float64} with indices SOneTo(3)×SOneTo(3)(0.61801, 0.53973, 0.77876):
  0.596987  -0.473236   0.647807
  0.772707   0.556304  -0.305697
 -0.215711   0.683062   0.697779

julia> log(exp(s))
3×3 InfinitesimalRotationVec{Float64} with indices SOneTo(3)×SOneTo(3)(0.61801, 0.53973, 0.77876):
  0.0      -0.77876   0.53973
  0.77876   0.0      -0.61801
 -0.53973   0.61801   0.0

julia> log(exp(s)) + s
3×3 InfinitesimalRotMatrix{3, Float64, 9} with indices SOneTo(3)×SOneTo(3):
  0.0      -1.55752   1.07946
  1.55752   0.0      -1.23602
 -1.07946   1.23602   0.0

julia> rotation_angle(exp(log(exp(s)) + s))
2.2624870696189996

julia> rotation_angle(exp(s))
1.1312435348094998

julia> InfinitesimalRotationVec(s)
3×3 InfinitesimalRotationVec{Float64} with indices SOneTo(3)×SOneTo(3)(0.61801, 0.53973, 0.77876):
  0.0      -0.77876   0.53973
  0.77876   0.0      -0.61801
 -0.53973   0.61801   0.0
```

### CODE REVIEW

**Review by andyferris — CHANGES_REQUESTED** [review]
Oh cool. :)

I would still avoid the term "infinitesimal rotation". Julia can represent infinitesimal rotations already, but it's not these. Here's one, with dual numbers:

```julia
using DualNumbers

infinitesimal_rotation = [
    1.0 Dual(0.0, 0.5);
    Dual(0.0, -0.5) 1.0
]
```

What you have in this PR is a rotation _generator_, so if we aren't going to refer to the matrix structure then we may as well call it by it's real name. Note the generator for the above an infinitesimal generator:

```julia
infinitesimal_generator = [
    0.0 Dual(0.0, 0.5);
    Dual(0.0, -0.5) 0.0
]
```

It's easy to figure this out using just `exp(ϵ) ≈ I + ϵ`.

What doesn't quite work out-of-the-box in Julia is `exp` and `log` with Dual matrices, but mathematically the following should hold:

```julia
exp(infinitesimal_generator) == infinitesimal_rotation
```

Hopefully with your work here we'll be able to get all these relationships working even with autodifferentiation on the parameters of the generator! :)

**Comment by hyrodium** [comment]
>I would still avoid the term "infinitesimal rotation". Julia can represent infinitesimal rotations already, but it's not these. Here's one, with dual numbers:

Ah, I didn't quite understand the meaning of inifinitesimal rotation.
Should we just rename the skew-symmetric matrix types like this?

* `InfinitesimalRotation` -> `RotationGenerator`
* `InfinitesimalRotMatrix` -> `RotMatrixGenerator`
* `InfinitesimalRotationVec` -> `RotationVecGenerator`
* `InfinitesimalAngle2c` -> `Angle2dGenerator`

>Hopefully with your work here we'll be able to get all these relationships working even with autodifferentiation on the parameters of the generator! :)

Yeah, I'll try to do that! :smile:

**Comment by andyferris** [comment]
> Should we just rename the skew-symmetric matrix types like this?

Yes - I think that is perfect :)

File: src/rotation_generator.jl [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+# RotationGenerator matrices should be skew-symmetric. Only the operations we define,
+# like addition, will stay as RotationGenerators, otherwise users will get an
+# SMatrix{3,3} (e.g. rot1 * rot2 -> SMatrix)
+Base.@pure StaticArrays.similar_type(::Union{R,Type{R}}) where {R <: RotationGenerator} = SMatrix{size(R)..., eltype(R), prod(size(R))}
+Base.@pure StaticArrays.similar_type(::Union{R,Type{R}}, ::Type{T}) where {R <: RotationGenerator, T} = SMatrix{size(R)..., T, prod(size(R))}
+
+# Useful for converting arrays of rotations to another rotation eltype, for instance.
+# Only works because parameters of all the rotations are of a similar form
+# Would need to be more sophisticated if we have arbitrary dimensions, etc
+@inline function Base.promote_op(::Type{R1}, ::Type{R2}) where {R1 <: RotationGenerator, R2 <: RotationGenerator}
```
Comment (andyferris) [inline src/rotation_generator.jl:43]: Is this right? The operator is not defined. Did you mean `promote_rule`?

**Review by andyferris — APPROVED** [review]
Awesome.

I had questions about the method defined for `promote_op` (and perhaps whatever you are trying to acheive there could be tested, too?) but otherwise this looks good to me.

This PR fills a massive gap, and I'm sure the community will appreciate it. Thanks @hyrodium!

File: src/rotation_generator.jl (hunk 1 above)
Comment (hyrodium) [reply src/rotation_generator.jl:43]: Ah, I was just referring to [`src/core_types.jl`](https://github.com/JuliaGeometry/Rotations.jl/blob/v1.1.0/src/core_types.jl#L68).
https://github.com/JuliaGeometry/Rotations.jl/commit/0ccc87629e00df8973d8d21cfc1994e3dfde1eb1#diff-55fa94d9eb6d180ed431c8aae0ad29c7100a0ad88cbd35dfc81239882a614e8dR39

Comment (hyrodium) [reply src/rotation_generator.jl]: I've tried renaming `promote_op` -> `promote_rule` and `isleaftype` -> `isconcretetype`, but it doesn't work well here.
These seems unnecessary, so I've dropped both `promote_op` definitions from `src/core_types.jl` and `src/rotation_generator.jl`.

**Pull request merged**